### PR TITLE
Make purification_type string and change numerical values

### PR DESCRIPTION
**Pull request by lmpawan10**
The following changes have been made:

**Inside** _**HardwareMonitor.h**_
- `purification_type` declared as string

**Inside** _**HardwareMonitor.cc**_
- `2002`: `"R1_Single_X-R1_Single_Z"`
- `3003`: `"R1_Single_X-R2_Single_Z"`
- `1001`: `"R1_Single_XZ"`
- `1221`: `"R1_Single_XZ-R2_Single_ZX"`
- `1011`: `"R1_Double_X"`
- `1021`: `"R1_Double_X-R2_Double_Z"`
- `1031`: `"R1_Double_XZ-R2_Double_ZX"`
- `1061`: `"R1_Double_X_Single_Z-R2_Double_Z_Single_X"`
- `5555`: `"CR1_Double_X-CR2_Double_ZX--R1_Single_X-R2_Single_Z"`
- `5556`: `"CR1_Double_X-R1_Single_Z-R2_Single_X"`

Here, R1/R2 is for specifying the number of rounds.
_i.e, R1: Round 1, R2: Round 2_

`"CR1_Double_X-CR2_Double_ZX--R1_Single_X-R2_Single_Z"` states that, 
Double_X and Double_ZX purification run in Round1 and Round2 respectively only once.
Then, Single_X and Single_Z purification run during Round1 and Round2 respectively and repeat for more rounds if needed.

- - -
This change is [Reviewable](https://reviewable.io/reviews/sfc-aqua/quisp/554)

### CODE REVIEW

File: quisp/modules/QRSA/HardwareMonitor/HardwareMonitor.cc [hunk 1]
```diff
@@ -611,8 +612,8 @@ void HardwareMonitor::sendLinkTomographyRuleSet(int my_address, int partner_addr
   std::vector<int> partners = {partner_address};
 
   if (num_purification > 0) {
-    if (purification_type == 2002) {  // Performs both X and Z purification for each n.
-      /// # Purification_type 2002: #
+    if (purification_type == "R1_Single_X-R1_Single_Z") {  // Performs both X and Z purification for each n.
```
Comment (zigen) [inline quisp/modules/QRSA/HardwareMonitor/HardwareMonitor.cc:615]: `R1_SINGLE_X_R1_SINGLE_Z`

* keep the naming convention: CAPITAL_CASE
* no dash `-`

**Review by Naphann — CHANGES_REQUESTED** [review]
I'd suggest that the general naming should give us the idea of the number of rounds as well. Say we put N rounds, some of the specify protocols run more than N rounds. 

Some ideas,
- 2002: `2N_ALTERNATE_SINGLE_X_WITH_SINGLE_Z`
- 3003: `N_ALTERNATE_SINGLE_X_WITH_SINGLE_Z`

It can get tricky with 1061, 5555, and 5556. Maybe something similar to below (still need to think of a way to specify it better)
- 1061: `N_ALTERNATE_DOUBLE_X_SINGLE_Z_WITH_DOUBLE_Z_SINGLE_X`
- 5555: `2+N_DOUBLE_X_THEN_DOUBLE_Z_ALTERNATE_SINGLE_X_WITH_SINGLE_Z`

Also it is possible that 5555 has wrong implementation, it shouldn't be DSDA or DSDA_INV since the comments said we want 3 qubits.

> Reviewed all commit messages.
> *[Reviewable](https://reviewable.io/reviews/sfc-aqua/quisp/554#-:-NyfUTF2-QcrHHMK2hHM:bk2zz69)* status: 0 of 27 files reviewed, 3 unresolved discussions (waiting on @lmpawan10)

___
*[`quisp/modules/QRSA/HardwareMonitor/HardwareMonitor.h` line 55 at r1](https://reviewable.io/reviews/sfc-aqua/quisp/554#-NyfTsE15CF1FTr9bl4C:-NyfTsE86EyqrhZDe2Zv:bwe72n2) ([raw file](https://github.com/sfc-aqua/quisp/blob/a18df66bf01e2f6070b84da006868806a86f2211/quisp/modules/QRSA/HardwareMonitor/HardwareMonitor.h#L55)):*
> ```objc
>   bool X_Purification = false;
>   bool Z_Purification = false;
>   // int purification_type = -1;
> ```

if we don't need it after the change, please delete it.
___
*[`quisp/modules/QRSA/HardwareMonitor/HardwareMonitor.cc` line 65 at r1](https://reviewable.io/reviews/sfc-aqua/quisp/554#-NyfUAdFD6mnZq0IxE6y:-NyfUAdFD6mnZq0IxE6z:b-8n6piq) ([raw file](https://github.com/sfc-aqua/quisp/blob/a18df66bf01e2f6070b84da006868806a86f2211/quisp/modules/QRSA/HardwareMonitor/HardwareMonitor.cc#L65)):*
> ```cpp
>   Z_Purification = par("z_purification");
>   purification_type = par("purification_type").str();
>   purification_type = purification_type.substr(1, purification_type.length() - 2);
> ```

why not use `stdstringValue()` of OMNeT?

**Comment by lmpawan10** [comment]
@Naphann what does 2 in 2002: `2N_ALTERNATE_SINGLE_X_WITH_SINGLE_Z` mean?

**Comment by lmpawan10** [comment]
For 5555, 
In the first round, I will change `DOUBLE_SELECTION_ZX_PURIFICATION` to `DOUBLE_SELECTION_Z_PURIFICATION`.

**Review by Naphann — COMMENTED** [review]
Hardware monitor takes in the purification type (currently the numerical value) and the number of rounds. For some types, the round input is not the actual message exchanging round. It is much clearer to the users and for semantics to make the number of purification rounds explicitly known from the type.

As we can see from the comments, `- rounds: ` that the input rounds get turned into sometimes `n`, sometimes `2n`, or sometimes `2+n`. 

> *[Reviewable](https://reviewable.io/reviews/sfc-aqua/quisp/554#-:-NysTut27gIl0yKKPbbI:b9zx9)* status: 0 of 27 files reviewed, 3 unresolved discussions (waiting on @lmpawan10)

**Review by lmpawan10 — COMMENTED** [review]
I got you. I have made the following changes accordingly.
-   2002: `2N_ALTERNATE_SINGLE_X_WITH_SINGLE_Z`
-   3003: `N_ALTERNATE_SINGLE_X_WITH_SINGLE_Z`
-   1001: `N_SINGLE_XZ`
-   1221: `N_ALTERNATE_SINGLE_XZ_WITH_SINGLE_ZX`
-   1011: `N_DOUBLE_X`
-   1021: `N_ALTERNATE_DOUBLE_X_WITH_DOUBLE_Z`
-   1031: `N_ALTERNATE_DOUBLE_XZ_WITH_DOUBLE_ZX`
-   1061: `N_ALTERNATE_DOUBLE_X_SINGLE_Z_WITH_DOUBLE_Z_SINGLE_X`
-   5555: `2+N_DOUBLE_X_THEN_DOUBLE_Z_ALTERNATE_SINGLE_X_WITH_SINGLE_Z`
-   5556: `1+N_DOUBLE_X_ALTERNATE_SINGLE_Z_WITH_SINGLE_X`

> *[Reviewable](https://reviewable.io/reviews/sfc-aqua/quisp/554#-:-NzZA-UKE_Y-9oADs8KV:b-hhlw53)* status: 0 of 29 files reviewed, 3 unresolved discussions

**Review by lmpawan10 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/sfc-aqua/quisp/554)* status: 0 of 29 files reviewed, 3 unresolved discussions (waiting on @Naphann)

___
*[`quisp/modules/QRSA/HardwareMonitor/HardwareMonitor.h` line 55 at r1](https://reviewable.io/reviews/sfc-aqua/quisp/554#-NyfTsE15CF1FTr9bl4C:-NzZFO5BCO0AYU7jLy2X:b-896fix) ([raw file](https://github.com/sfc-aqua/quisp/blob/a18df66bf01e2f6070b84da006868806a86f2211/quisp/modules/QRSA/HardwareMonitor/HardwareMonitor.h#L55)):*
<details><summary><i>Previously, Naphann (Naphan Benchasattabuse) wrote…</i></summary><blockquote>

if we don't need it after the change, please delete it.
</blockquote></details>

Done.

**Review by lmpawan10 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/sfc-aqua/quisp/554)* status: 0 of 29 files reviewed, 3 unresolved discussions (waiting on @lmpawan10, @Naphann, and @zigen)

___
*[`quisp/modules/QRSA/HardwareMonitor/HardwareMonitor.cc` line 615 at r1](https://reviewable.io/reviews/sfc-aqua/quisp/554#gh-1613136754:-NzZFWepET4z7ozE_zdN:b-896fix) ([raw file](https://github.com/sfc-aqua/quisp/blob/a18df66bf01e2f6070b84da006868806a86f2211/quisp/modules/QRSA/HardwareMonitor/HardwareMonitor.cc#L615)):*
<details><summary><i>Previously, zigen (Kentaro "zigen" Teramoto) wrote…</i></summary><blockquote>

`R1_SINGLE_X_R1_SINGLE_Z`

* keep the naming convention: CAPITAL_CASE
* no dash `-` 
</blockquote></details>

Done.

**Review by Naphann — CHANGES_REQUESTED** [review]
> Reviewed 2 of 27 files at r1, 27 of 27 files at r2.
> *[Reviewable](https://reviewable.io/reviews/sfc-aqua/quisp/554)* status: all files reviewed (commit messages unreviewed), 2 unresolved discussions (waiting on @lmpawan10 and @zigen)

___
*[`quisp/modules/QRSA/HardwareMonitor/HardwareMonitor.cc` line 65 at r1](https://reviewable.io/reviews/sfc-aqua/quisp/554#-NyfUAdFD6mnZq0IxE6y:-NzludP2DQupJPvg1xqg:b-igrlaq) ([raw file](https://github.com/sfc-aqua/quisp/blob/a18df66bf01e2f6070b84da006868806a86f2211/quisp/modules/QRSA/HardwareMonitor/HardwareMonitor.cc#L65)):*
<details><summary><i>Previously, lmpawan10 (Pawan Poudel) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

I was thinking of this one https://doc.omnetpp.org/omnetpp4/api/classcPar.html#aea60efd9c5f66088badcf921c8decd68. It would look a little cleaner if it works.

**Review by lmpawan10 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/sfc-aqua/quisp/554)* status: 28 of 29 files reviewed, 2 unresolved discussions (waiting on @Naphann and @zigen)

___
*[`quisp/modules/QRSA/HardwareMonitor/HardwareMonitor.cc` line 65 at r1](https://reviewable.io/reviews/sfc-aqua/quisp/554#-NyfUAdFD6mnZq0IxE6y:-Nzqwhcx2PQPeFNT162Y:b-896fix) ([raw file](https://github.com/sfc-aqua/quisp/blob/a18df66bf01e2f6070b84da006868806a86f2211/quisp/modules/QRSA/HardwareMonitor/HardwareMonitor.cc#L65)):*
<details><summary><i>Previously, Naphann (Naphan Benchasattabuse) wrote…</i></summary><blockquote>

I was thinking of this one https://doc.omnetpp.org/omnetpp4/api/classcPar.html#aea60efd9c5f66088badcf921c8decd68. It would look a little cleaner if it works.
</blockquote></details>

Done.

**Review by Naphann — APPROVED** [review]
:lgtm: 

> Reviewed 1 of 1 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/sfc-aqua/quisp/554#-:-NzrGpzI7U4U3C01J5mE:bg7igkv)* status: all files reviewed, 1 unresolved discussion (waiting on @zigen)

**Review by zigen — APPROVED** [review]
looks much better than before! thanks! :lgtm:

> Reviewed 1 of 1 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/sfc-aqua/quisp/554#-:-O11OSXt3tZ39iShsuyk:b-ku6e8t)* status: <span>:shipit:</span> complete! all files reviewed, all discussions resolved (waiting on @lmpawan10)

**Pull request merged**